pkg/registry/apps/controllerrevision/storage/storage.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package storage
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/generic"
+	genericregistry "k8s.io/apiserver/pkg/registry/generic/registry"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/apis/apps"
+	"k8s.io/kubernetes/pkg/registry/apps/controllerrevision"
+	"k8s.io/kubernetes/pkg/registry/cachesize"
+)
+
+// REST implements a RESTStorage for AnyState
+type REST struct {
+	*genericregistry.Store
+}
+
+// NewREST returns a RESTStorage object that will work with AnyState objects.
+func NewREST(optsGetter generic.RESTOptionsGetter) *REST {
+	store := &genericregistry.Store{
+		Copier:            api.Scheme,
+		NewFunc:           func() runtime.Object { return &apps.ControllerRevision{} },
+		NewListFunc:       func() runtime.Object { return &apps.ControllerRevisionList{} },
+		PredicateFunc:     controllerrevision.MatchControllerRevision,
+		QualifiedResource: apps.Resource("controllerrevisions"),
+		WatchCacheSize:    cachesize.GetWatchCacheSizeByResource("controllerrevisions"),
+
+		CreateStrategy: controllerrevision.Strategy,
+		UpdateStrategy: controllerrevision.Strategy,
+		DeleteStrategy: controllerrevision.Strategy,
+	}
+	options := &generic.StoreOptions{RESTOptions: optsGetter, AttrFunc: controllerrevision.GetAttrs}
+	if err := store.CompleteWithOptions(options); err != nil {
+		panic(err)
+	}
+	return &REST{store}
+}
+
+// Implement ShortNamesProvider
+var _ rest.ShortNamesProvider = &REST{}
+
+// ShortNames implements the ShortNamesProvider interface. Returns a list of short names for a resource.
+func (r *REST) ShortNames() []string {
+	return []string{"cr"}
```

Comment:
@deads2k now that short names are not consolidated I wonder how can we avoid collisions with other short names. What happens if someone checks in a short name that already exists? Do we have a test to prevent such a scenario?